chainimport/headers_import.go

```diff
@@ -0,0 +1,1814 @@
+	headerFactory func() Header
+}
+
+// Compile-time assertion to ensure FileHeaderImportSource implements
+// HeaderImportSource interface.
+var _ HeaderImportSource = (*FileHeaderImportSource)(nil)
+
+// Open opens the file and initializes the mmap reader.
+func (f *FileHeaderImportSource) Open() error {
+	reader, err := mmap.Open(f.GetURI())
+	if err != nil {
+		return fmt.Errorf("failed to mmap file: %w", err)
+	}
+	f.reader = reader
+	f.fileSize = reader.Len()
+
+	// Read and initialize metadata from the beginning of the file.
+	mData, err := f.GetHeaderMetadata()
+	if err != nil {
+		return fmt.Errorf("failed to get header metadata: %w", err)
+	}
+	f.metadata = mData
+	f.metadata.EndHeight = mData.StartHeight + mData.HeadersCount - 1
+
+	return nil
+}
+
+// Close closes the file and releases the mmap reader.
+func (f *FileHeaderImportSource) Close() error {
+	return f.reader.Close()
+}
+
+// GetHeaderMetadata reads the metadata from the file. The metadata is memoized
+// after the first call, with subsequent calls returning the cached result
+// without re-reading the file.
+func (f *FileHeaderImportSource) GetHeaderMetadata() (*HeaderMetadata, error) {
+	if f.metadata != nil {
+		return f.metadata, nil
+	}
+
+	if f.reader == nil {
+		return nil, errors.New("file reader not initialized")
+	}
+
+	// Read the first HeaderMetadataSize bytes containing metadata.
+	buf := make([]byte, HeaderMetadataSize)
+	_, err := f.reader.ReadAt(buf, 0)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read metadata: %w", err)
+	}
+
+	// Calculate offsets.
+	headerTypeOffset := BitcoinChainTypeSize
+	startHeightOffset := BitcoinChainTypeSize + HeaderTypeSize
+
+	// Extract the bitcoin chain type.
+	bitcoinChainType := wire.BitcoinNet(
+		binary.LittleEndian.Uint32(buf[:BitcoinChainTypeSize]),
+	)
+
+	// Extract the header type
+	headerType := headerfs.HeaderType(buf[headerTypeOffset])
+
+	// Extract start height. First check if the value is negative.
+	startHeightRaw := buf[startHeightOffset:HeaderMetadataSize]
+	var startHeight int32
+	if err = binary.Read(
+		bytes.NewReader(startHeightRaw), binary.LittleEndian,
+		&startHeight,
+	); err != nil {
+		return nil, fmt.Errorf("failed to read start height: %w", err)
+	}
+
+	if startHeight < 0 {
+		return nil, fmt.Errorf("invalid negative value detected for "+
+			"StartHeight: %d", startHeight)
+	}
+
+	// Construct the header metadata.
+	metadata := HeaderMetadata{
+		BitcoinChainType: bitcoinChainType,
+		HeaderType:       headerType,
+		StartHeight:      uint32(startHeight),
+	}
+
+	// Construct header type size. This also validates the header type. If
+	// the header type is not valid, we will get an error while getting the
+	// size of this header type.
+	headerSize, err := metadata.HeaderType.Size()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get header size: %v", err)
+	}
+	metadata.HeaderSize = headerSize
+
+	// Compute headers count.
+	usableFileSize := f.fileSize - HeaderMetadataSize
```

Comment:
Should we check that this is actually greater than zero here? Meaning that no empty file is provided (just a header but no content)?